File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
from  collections import  deque

T = int(sys.stdin.readline().rstrip())


def bfs(x, y, graph, N, M):
    graph[x][y] = 2
    directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

    queue = deque()
    queue.append([x,y])

    while queue:
        cur_x, cur_y = queue.popleft()

        for i in range(4):
            dx, dy = directions[i][0], directions[i][1]
            nx, ny = cur_x + dx, cur_y + dy
            if 0 <= nx < N and 0 <= ny < M and graph[nx][ny] == 1:
                graph[nx][ny] = 2
                queue.append([nx,ny])


for _ in range(T):
    M, N, K = map(int, sys.stdin.readline().split())
    graph = [[0] * M for _ in range(N)]
    count = 0
    for _ in range(K):
        x, y = map(int, sys.stdin.readline().split())
        graph[y][x] = 1

    for i in range(N):
        for j in range(M):
            if graph[i][j] == 1:
                bfs(i, j, graph, N, M)
                count += 1
    print(count)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.util.*;
import java.io.*;
class Main {
    static int N, M;
    static int[] dx = new int[]{0, 0, -1, 1};
    static int[] dy = new int[]{-1, 1, 0, 0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
          String[] row = br.readLine().split(" ");
          N = Integer.parseInt(row[0]);
          M = Integer.parseInt(row[1]);

          int[][] graph = new int[N][M];

          int K = Integer.parseInt(row[2]);

          for (int i = 0; i < K; i++) {
            String[] info = br.readLine().split(" ");

            int x = Integer.parseInt(info[0]);
            int y = Integer.parseInt(info[1]);

            graph[x][y] = 1;

          }

          boolean[][] visited = new boolean[N][M];
          Deque<int[]> queue = new ArrayDeque<>();
          int cnt = 0;
          for (int i =0; i < N; i++) {
            for (int j = 0; j < M; j++) {
              if (!visited[i][j] && graph[i][j] == 1) {
                bfs(i, j, graph, visited, queue);
                cnt++;
              }
            }
          }

          System.out.println(cnt);

     
        }

    }
    static void bfs(int x, int y, int[][] graph, boolean[][] visited, Deque<int[]> queue ) {
      queue.add(new int[]{x, y});
      visited[x][y] = true;

      while (!queue.isEmpty()) {

        int[] curr = queue.poll();

        int cx = curr[0]; int cy = curr[1];

        for (int d = 0 ;d < 4; d++) {
          int nx = cx + dx[d]; int ny = cy + dy[d];

          if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
          if (!visited[nx][ny] && graph[nx][ny] == 1) {
            visited[nx][ny] = true;
            queue.add(new int[]{nx, ny});
          }
        }
        
      }
    }
}